aidial_adapter_openai/gpt4_vision/chat_completion.py

```diff
@@ -0,0 +1,374 @@
+        for attachment in attachments
+    ]
+
+    logger.debug(f"download results: {download_results}")
+
+    errors: List[Tuple[int, str]] = [
+        (idx, result)
+        for idx, result in enumerate(download_results)
+        if isinstance(result, str)
+    ]
+
+    if len(errors) > 0:
+        logger.debug(f"download errors: {errors}")
+        return errors
+
+    image_urls: List[ImageDataURL] = cast(List[ImageDataURL], download_results)
+
+    image_tokens: List[int] = []
+    image_messages: List[dict] = []
+
+    for image_url in image_urls:
+        tokens, detail = tokenize_image(image_url, "auto")
+        image_tokens.append(tokens)
+        image_messages.append(create_image_message(image_url, detail))
+
+    total_image_tokens = sum(image_tokens)
+
+    logger.debug(f"image tokens: {image_tokens}")
+
+    sub_messages: List[dict] = [create_text_message(content)] + image_messages
+
+    return {**message, "content": sub_messages}, total_image_tokens
+
+
+async def transform_messages(
+    file_storage: Optional[FileStorage], messages: List[dict]
+) -> Tuple[List[dict], int] | str:
+    new_messages: List[dict] = []
+    image_tokens = 0
+
+    errors: Dict[int, List[Tuple[int, str]]] = {}
+
+    n = len(messages)
+    for idx, message in enumerate(messages):
+        result = await transform_message(file_storage, message)
+        if isinstance(result, list):
+            errors[n - idx] = result
+        else:
+            new_message, tokens = result
+            new_messages.append(new_message)
+            image_tokens += tokens
+
+    if errors:
+        msg = "Some of the image attachments failed to download:"
+        for i, error in errors.items():
+            msg += f"\n- {format_ordinal(i)} message from end:"
+            for j, err in error:
+                msg += f"\n  - {format_ordinal(j + 1)} attachment: {err}"
+        return msg
+
+    return new_messages, image_tokens
+
+
+T = TypeVar("T")
+
+
+async def chat_completion(
+    request: Any,
+    deployment: str,
+    upstream_endpoint: str,
+    api_key: str,
+    is_stream: bool,
+    file_storage: Optional[FileStorage],
+) -> Response:
+    if request.get("n", 1) > 1:
+        raise HTTPException(
+            status_code=422,
+            message="The deployment doesn't support n > 1",
+            type="invalid_request_error",
+        )
+
+    messages: List[Any] = request["messages"]
+    if len(messages) == 0:
+        raise HTTPException(
+            status_code=422,
+            message="The request doesn't contain any messages",
+            type="invalid_request_error",
+        )
+
+    api_url = upstream_endpoint + "?api-version=2023-12-01-preview"
+
+    result = await transform_messages(file_storage, messages)
+
+    if isinstance(result, str):
+        logger.debug(f"Failed to prepare request for GPT4V: {result}")
+
+        if file_storage is not None:
+            # Report usage-level error if the request came from the chat
```

Comment:
What does usage-level mean?

Reply:
Fixed. It's a typo - I meant `user-level error` (as opposed to `internal error` which chat renders as red boxes).